### PR TITLE
ft: support specifying partition in entries

### PR DESCRIPTION
**Pull request by philipyoo**
To support deep healthcheck, where I want to produce to all
partitions, allow for specifying a partition within an entry.

### CODE REVIEW

File: lib/BackbeatProducer.js [hunk 1]
```diff
@@ -128,17 +128,28 @@ class BackbeatProducer extends EventEmitter {
...
-                  }));
+
+        const payload = entries.map(item => {
+            if (this._partition === undefined && item.partition === undefined) {
+                return {
+                    topic: this._topic,
+                    messages: new KeyedMessage(item.key, item.message),
+                    key: item.key,
+                };
+            } else if (item.partition && this._partition === undefined) {
```
Comment (jonathan-gramain) [inline lib/BackbeatProducer.js]: You can get rid of the `else`.

Minor style comment: the tests would be simpler and more understandable if you use the following sequence instead:

```
if (this._partition) { return ...}
if (item.partition) { return ...}
return {...}
```

**Review by jonathan-gramain — DISMISSED** [review]

File: lib/BackbeatProducer.js [hunk 2]
```diff
@@ -128,17 +128,29 @@ class BackbeatProducer extends EventEmitter {
...
-                      key: item.key,
-                  })) :
-                  entries.map(item => ({
-                      topic: this._topic,
-                      messages: item.message,
-                      partition: this._partition,
-                  }));
+
+        const payload = entries.map(item => {
+            if (item.partition) {
```
Comment (rahulreddy) [inline lib/BackbeatProducer.js]: This needs to be `if (item.partition !== undefined)` since partition can be `0`.

File: lib/BackbeatProducer.js [hunk 3]
```diff
@@ -128,17 +128,29 @@ class BackbeatProducer extends EventEmitter {
...
+                return {
+                    topic: this._topic,
+                    messages: item.message,
+                    partition: item.partition,
+                };
+            }
+            if (this._partition) {
```
Comment (rahulreddy) [inline lib/BackbeatProducer.js]: Same here `if (this._partition !== undefined)`

**Review by rahulreddy — CHANGES_REQUESTED** [review]

**Comment by philipyoo** [comment]
Sorry, one additional field I'd like to add for entries that have explicit partition declared. Added PR: 79de399

I realize when parsing entries in `QueueProcessor.processKafkaEntry`, I want to catch entries specific to deep healthcheck. By setting a `key` property for these entries, I can catch and ignore deep healthcheck entries.

**Comment by jonathan-gramain** [comment]
@philipyoo I had a similar issue with testing to inject artificial messages and ensure the consumer would receive them, while not disturbing the regular backbeat consumer too much. For this I added a "bootstrapId" field to the JSON message, and backbeat consumer skips those entries (actually sends back an error but keeps going with next entries).

In case of healthchecks, we do not want to error but simply skip them in regular processing. I'm thinking `key` can be confused with other fields with the same name (object key, kafka key...), hence maybe just setting a `"healthcheck": true` or `"healthcheckKey": "foo"` field can do the trick, but make sure that it's clearly visible somehow that the entry comes from the healthcheck to ease troubleshooting.

**Comment by philipyoo** [comment]
As @jonathan-gramain commented, rather than catching deep healthcheck entries by the key property, instead putting the identifier within the message property and filtering within `QueueEntry.createFromKafkaEntry`

**Comment by philipyoo** [comment]
Implementation has changed with switch to node-rdkafka. Plan is now to produce to Kafka directly using the same topic and consumer group.

**Pull request closed without merging**